src/eg/src/serialize.rs

```diff
@@ -0,0 +1,53 @@
+use anyhow::{Context, Result};
+use std::io::Cursor;
+
+pub trait SerializableCanonical {
+    /// Writes an entity to a [`std::io::Write`] as canonical bytes.
+    /// This uses a more compact JSON format.
+    fn to_stdiowrite_canonical(&self, stdiowrite: &mut dyn std::io::Write) -> Result<()>
+    where
+        Self: serde::Serialize,
+    {
+        serde_json::ser::to_writer(stdiowrite, self).context("Writing canonical")
+    }
+
+    /// Returns the canonical byte sequence representation of the entity.
+    /// This uses a more compact JSON format.
+    fn to_canonical_bytes(&self) -> Result<Vec<u8>>
+    where
+        Self: serde::Serialize,
+    {
+        let mut buf = Cursor::new(Vec::new());
+        self.to_stdiowrite_canonical(&mut buf)
+            .context("Writing canonical")?;
+        Ok(buf.into_inner())
+    }
+}
+
+pub trait SerializablePretty {
+    /// Returns a pretty JSON `String` representation of the entity.
+    /// The final line will end with a newline.
+    fn to_json_pretty(&self) -> String
+    where
+        Self: serde::Serialize,
+    {
+        // `unwrap()` is justified here because why would JSON serialization fail?
+        #[allow(clippy::unwrap_used)]
+        let mut s = serde_json::to_string_pretty(self).unwrap();
+        s.push('\n');
+        s
+    }
+
+    /// Writes an entity to a [`std::io::Write`] as pretty JSON.
+    fn to_stdiowrite(&self, stdiowrite: &mut dyn std::io::Write) -> Result<()>
```

Comment:
Can we please call this `to_stdiowrite_pretty()` so it's clear at the point-of-call?

Reply:
Also, I know this is silly and meaningless, but can we please move `to_stdiowrite()` above `to_json_pretty()` in the order in the trait to match the logical order of the other trait?

Reply:
> Also, I know this is silly and meaningless,

Nah, I like the attention to detail.